source/extensions/common/tap/admin.cc

```diff
@@ -0,0 +1,110 @@
+      SINGLETON_MANAGER_REGISTERED_NAME(tap_admin_handler), [&admin, &main_thread_dispatcher] {
+        return std::make_shared<AdminHandler>(admin, main_thread_dispatcher);
+      });
+}
+
+AdminHandler::AdminHandler(Server::Admin& admin, Event::Dispatcher& main_thread_dispatcher)
+    : admin_(admin), main_thread_dispatcher_(main_thread_dispatcher) {
+  bool rc =
+      admin_.addHandler("/tap", "tap filter control", MAKE_ADMIN_HANDLER(handler), true, true);
+  RELEASE_ASSERT(rc, "/tap admin endpoint is taken");
+}
+
+AdminHandler::~AdminHandler() {
+  bool rc = admin_.removeHandler("/tap");
+  ASSERT(rc);
+}
+
+Http::Code AdminHandler::handler(absl::string_view, Http::HeaderMap&, Buffer::Instance& response,
+                                 Server::AdminStream& admin_stream) {
+  if (attached_request_.has_value()) {
+    response.add("An attached /tap admin stream already exists. Detach it.");
+    return Http::Code::BadRequest;
+  }
+
+  if (admin_stream.getRequestBody() == nullptr) {
+    response.add("/tap requires a JSON/YAML body");
+    return Http::Code::BadRequest;
+  }
+
+  envoy::admin::v2alpha::TapRequest tap_request;
+  try {
+    MessageUtil::loadFromYamlAndValidate(admin_stream.getRequestBody()->toString(), tap_request);
+  } catch (EnvoyException& e) {
+    response.add(e.what());
+    return Http::Code::BadRequest;
+  }
+
+  ENVOY_LOG(debug, "tap admin request for config_id={}", tap_request.config_id());
+  if (config_id_map_.count(tap_request.config_id()) == 0) {
+    response.add(fmt::format("Unknown config id '{}'. No extension has registered with this id.",
+                             tap_request.config_id()));
+    return Http::Code::BadRequest;
+  }
+  for (auto config : config_id_map_[tap_request.config_id()]) {
+    config->newTapConfig(std::move(*tap_request.mutable_tap_config()), this);
+  }
+
+  admin_stream.setEndStreamOnComplete(false);
+  admin_stream.addOnDestroyCallback([this] {
+    for (auto config : config_id_map_[attached_request_.value().config_id_]) {
+      ENVOY_LOG(debug, "detach tap admin request for config_id={}",
+                attached_request_.value().config_id_);
+      config->clearTapConfig();
+      attached_request_ = absl::nullopt;
+    }
+  });
+  attached_request_.emplace(tap_request.config_id(), &admin_stream);
+  return Http::Code::OK;
+}
+
+void AdminHandler::registerConfig(ExtensionConfig& config, const std::string& config_id) {
+  ASSERT(!config_id.empty());
+  ASSERT(config_id_map_[config_id].count(&config) == 0);
+  config_id_map_[config_id].insert(&config);
+}
+
+void AdminHandler::unregisterConfig(ExtensionConfig& config) {
+  ASSERT(!config.adminId().empty());
+  ASSERT(config_id_map_[config.adminId()].count(&config) == 1);
+  config_id_map_[config.adminId()].erase(&config);
+  if (config_id_map_[config.adminId()].empty()) {
+    config_id_map_.erase(config.adminId());
+  }
+}
+
+void AdminHandler::submitBufferedTrace(std::shared_ptr<Protobuf::Message> trace) {
```

Comment:
Do you want stronger typing on the trace here? The code works with `Message`, but realistically we should be using a more specific type.

Reply:
My thinking here is this is a generic interface that works regardless of the message type. When I port the capture transport socket over, this will allow a sink (admin, file, etc.) to implement a sink without knowing about the contained messages.

Reply:
I think for gRPC we're going to have to be more concrete though. Would we be sending `Any` objects out on the stream? Might be worth a few mins mocking up a gRPC service for this to see..

Reply:
Not sure using `Any` would help, if we send different types in different endpoint (gRPC method) then it is OK, otherwise I would define a message with oneof all possible types.

Reply:
Yeah I like Lizan's idea of having a wrapper message with a oneof inside of it. Let me try that out.